### PR TITLE
ARROW-6310: [C++] IPC json should use strings for 64 bit ints

### PR DESCRIPTION
**Pull request by bkietz**
JS can't represent all 64 bit integers as numbers (which are doubles), so store them in the JSON IPC as strings instead.

### CODE REVIEW

**Comment by pitrou** [comment]
You'll probably need to change the Java side as well (and perhaps Go). Maybe @emkornfield can help.

File: integration/integration_test.py [hunk 1]
```diff
@@ -184,6 +184,15 @@ def _get_buffers(self):
 
 
 class IntegerType(PrimitiveType):
+    class Column(PrimitiveColumn):
```
Comment (wesm) [inline integration/integration_test.py]: It isn't very idiomatic to declare nested classes, this should be `IntegerColumn` at the top level

File: cpp/src/arrow/type_traits.h [hunk 2]
```diff
@@ -441,6 +446,16 @@ struct is_8bit_int {
       (std::is_same<UInt8Type, T>::value || std::is_same<Int8Type, T>::value);
 };
 
+template <typename T, typename CType = void>
```
Comment (wesm) [inline cpp/src/arrow/type_traits.h]: Should put this in json_internal.cc unless we anticipate using it elsewhere

**Review by wesm — COMMENTED** [review]
We might need to set up an integration branch for this work

**Comment by wesm** [comment]
I set up https://issues.apache.org/jira/browse/ARROW-6512

**Comment by bkietz** [comment]
Java side seems to have been addressed already: https://github.com/apache/arrow/pull/5002

File: integration/integration_test.py (hunk 1 above)
Comment (bkietz) [reply integration/integration_test.py]: will do

File: cpp/src/arrow/type_traits.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/type_traits.h]: I do think `IsSignedInt` etc should be removed in favor of this trait or one like it (though it should probably be renamed).

`IsSignedInt` takes either an array or a type as a type argument, which is surprisingly atypical for traits. Furthermore whereas `is_signed_integer` returns false for date and other types which are represented by but not identical to integers `IsSignedInt` returns true by checking only the `c_type`, which leads to `static_assert(IsSignedInt<HalfFloatType>::value, "")`. Finally the declaration of IsSignedInt is far from readable due to nested macro usage

Comment (bkietz) [reply cpp/src/arrow/type_traits.h]: https://issues.apache.org/jira/browse/ARROW-6515

**Review by wesm — APPROVED** [review]
+1. I changed the base to an integration branch where we can address the Go and JS side of this before merging into master

**Comment by wesm** [comment]
See failure here https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/33330202/job/db2r0oruyib74qy6

and fix https://github.com/apache/arrow/pull/7352/commits/ae09c0dfe701ae10bba6251aced27b25e12608b2

**Pull request closed without merging**